src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs

```diff
@@ -438,21 +440,60 @@ private async Task<Project> UpdateProjectInfoAsync(Project project, Checksum inf
                 Func<Solution, ImmutableArray<DocumentId>, Solution> removeDocuments,
                 CancellationToken cancellationToken) where TDocumentState : TextDocumentState
             {
-                using var _1 = PooledHashSet<Checksum>.GetInstance(out var olds);
-                using var _2 = PooledHashSet<Checksum>.GetInstance(out var news);
+                using var _1 = PooledDictionary<DocumentId, Checksum>.GetInstance(out var oldDocumentIdToChecksum);
+                using var _2 = PooledDictionary<DocumentId, Checksum>.GetInstance(out var newDocumentIdToChecksum);
+                using var _3 = PooledHashSet<DocumentId>.GetInstance(out var allDocumentIds);
+
+                foreach (var (oldChecksum, documentId) in oldChecksums)
+                    oldDocumentIdToChecksum.Add(documentId, oldChecksum);
+
+                foreach (var (newChecksum, documentId) in newChecksums)
+                    newDocumentIdToChecksum.Add(documentId, newChecksum);
 
-                olds.AddRange(oldChecksums.Checksums.Children);
-                news.AddRange(newChecksums.Checksums.Children);
+                allDocumentIds.AddRange(oldChecksums.Ids);
+                allDocumentIds.AddRange(newChecksums.Ids);
 
-                // remove documents that exist in both side
-                olds.ExceptWith(newChecksums.Checksums);
-                news.ExceptWith(oldChecksums.Checksums);
+                // remove documents that are the same on both sides
+                foreach (var documentId in allDocumentIds)
```

Comment:
>foreach (var documentId in allDocumentIds) [](http://example.com/codeflow?start=16&length=42)

Can you just enumerate over oldChecksums and not need allDocumentIds?

Reply:
yup.  done.